rasterio/rio/merge.py

```diff
@@ -43,148 +43,27 @@ def merge(ctx, files, output, driver, bounds, res, nodata, creation_options):
-                        "range of its data type, %s. Consider overriding it "
-                        "using the --nodata option for better results." % (
-                            nodataval, dtype))
-            else:
-                nodataval = 0
-
-            dst_w, dst_s, dst_e, dst_n = dst.bounds
-
-            for fname in reversed(files):
-                with rasterio.open(fname) as src:
-                    # Real World (tm) use of boundless reads.
-                    # This approach uses the maximum amount of memory to solve
-                    # the problem. Making it more efficient is a TODO.
-
-                    # 1. Compute spatial intersection of destination
-                    #    and source.
-                    src_w, src_s, src_e, src_n = src.bounds
-
-                    int_w = src_w if src_w > dst_w else dst_w
-                    int_s = src_s if src_s > dst_s else dst_s
-                    int_e = src_e if src_e < dst_e else dst_e
-                    int_n = src_n if src_n < dst_n else dst_n
-
-                    # 2. Compute the source window.
-                    src_window = src.window(int_w, int_s, int_e, int_n)
-
-                    # 3. Compute the destination window.
-                    dst_window = dst.window(int_w, int_s, int_e, int_n)
-
-                    # 4. Initialize temp array.
-                    temp = np.zeros(
-                            (first.count,) + tuple(b - a for a, b in dst_window),
-                            dtype=dtype)
-
-                    temp = src.read(
-                            out=temp,
-                            window=src_window,
-                            boundless=False,
-                            masked=True)
-
-                    # 5. Copy elements of temp into dest.
-                    roff, coff = dst.index(int_w, int_n)
-                    h, w = temp.shape[-2:]
-
-                    region = dest[:,roff:roff+h,coff:coff+w]
-                    np.copyto(region, temp,
-                        where=np.logical_and(
-                        region==nodataval, temp.mask==False))
-
-            if dst.mode == 'r+':
-                temp = dst.read(masked=True)
-                np.copyto(dest, temp,
-                    where=np.logical_and(
-                    dest==nodataval, temp.mask==False))
-
-            dst.write(dest)
-            dst.close()
-
-    except Exception:
+        sources = [rasterio.open(f) for f in files]
+        dest, output_transform = merge_tool(sources, bounds=bounds, res=res,
+                                            nodata=nodata)
+        kwargs = sources[0].meta
+        kwargs.update(**creation_options)
+        kwargs.pop('affine')
+        kwargs['transform'] = output_transform
+        kwargs['height'] = dest.shape[1]
+        kwargs['width'] = dest.shape[2]
+        kwargs['driver'] = driver
+
+        dst = rasterio.open(output, 'w', **kwargs)
```

Comment:
Here let's use the newer `.profile` property instead of `.meta` following the example in rio-convert: https://github.com/mapbox/rasterio/blob/master/rasterio/rio/convert.py#L161-L180.